bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -3492,12 +3492,18 @@ void ReplicaImp::executeRequestsInPrePrepareMsg(concordUtils::SpanWrapper &paren
       ConcordAssertGT(actualReplyLength,
                       0);  // TODO(GG): TBD - how do we want to support empty replies? (actualReplyLength==0)
 
-      ClientReplyMsg *replyMsg = clientsManager->allocateNewReplyMsgAndWriteToStorage(
-          clientId, req.requestSeqNum(), currentPrimary(), replyBuffer, actualReplyLength);
-      replyMsg->setReplicaSpecificInfoLength(actualReplicaSpecificInfoLength);
-      send(replyMsg, clientId);
-      delete replyMsg;
-      clientsManager->removePendingRequestOfClient(clientId);
+      if (status != 0) {
+        const auto requestSeqNum = req.requestSeqNum();
+        LOG_DEBUG(CNSUS, "Request execution failed: " << KVLOG(clientId, requestSeqNum));
+        return;
+      } else {
+        ClientReplyMsg *replyMsg = clientsManager->allocateNewReplyMsgAndWriteToStorage(
+            clientId, req.requestSeqNum(), currentPrimary(), replyBuffer, actualReplyLength);
+        replyMsg->setReplicaSpecificInfoLength(actualReplicaSpecificInfoLength);
+        send(replyMsg, clientId);
+        delete replyMsg;
+        clientsManager->removePendingRequestOfClient(clientId);
```

Comment:
This should be called in both cases, otherwise a client retry will be ignored.

Reply:
Thanks, great catch!